### PR TITLE
Replace unnecessary synchronous or concurrent containers with ordinary containers

### PR DESCRIPTION
**Pull request by BurningCN**
## Checklist
- [x] Make sure there is a [GitHub_issue](https://github.com/apache/dubbo/issues) field for the change (usually before you start working on it). Trivial changes like typos do not require a GitHub issue. Your pull request should address just this issue, without pulling in other changes - one PR resolves one issue.
- [ ] Each commit in the pull request should have a meaningful subject line and body.
- [ ] Write a pull request description that is detailed enough to understand what the pull request does, how, and why.
- [ ] Check if is necessary to patch to Dubbo 3 if you are work on Dubbo 2.7
- [ ] Write necessary unit-test to verify your logic correction, more mock a little better when cross module dependency exist. If the new feature or significant change is committed, please remember to add sample in [dubbo samples](https://github.com/apache/dubbo-samples) project.
- [ ] Add some description to [dubbo-website](https://github.com/apache/dubbo-website) project if you are requesting to add a feature.
- [ ] GitHub Actions works fine on your own branch.
- [ ] If this contribution is large, please follow the [Software Donation Guide](https://github.com/apache/dubbo/wiki/Software-donation-guide).

### CODE REVIEW

File: dubbo-registry/dubbo-registry-api/src/main/java/org/apache/dubbo/registry/client/event/listener/ServiceInstancesChangedListener.java [hunk 1]
```diff
@@ -372,21 +370,20 @@ protected void notifyAddressChanged() {
      * Since this listener is shared among interfaces, destroy this listener only when all interface listener are unsubscribed
      */
     public synchronized void destroy() {
-        if (!destroyed.get()) {
+        if (!destroyed && CollectionUtils.isEmptyMap(listeners)) {
             if (CollectionUtils.isEmptyMap(listeners)) {
```
Comment (AlbumenJ) [inline dubbo-registry/dubbo-registry-api/src/main/java/org/apache/dubbo/registry/client/event/listener/ServiceInstancesChangedListener.java:374]: `CollectionUtils.isEmptyMap(listeners)` is roundabout

**Review by AlbumenJ — CHANGES_REQUESTED** [review]

File: dubbo-registry/dubbo-registry-api/src/main/java/org/apache/dubbo/registry/client/event/listener/ServiceInstancesChangedListener.java (hunk 1 above)
Comment (BurningCN) [reply dubbo-registry/dubbo-registry-api/src/main/java/org/apache/dubbo/registry/client/event/listener/ServiceInstancesChangedListener.java]: done

File: dubbo-registry/dubbo-registry-api/src/main/java/org/apache/dubbo/registry/client/event/listener/ServiceInstancesChangedListener.java [hunk 2]
```diff
@@ -84,8 +82,8 @@
     public ServiceInstancesChangedListener(Set<String> serviceNames, ServiceDiscovery serviceDiscovery) {
         this.serviceNames = serviceNames;
         this.serviceDiscovery = serviceDiscovery;
-        this.listeners = new ConcurrentHashMap<>();
-        this.listenerQueue = new ConcurrentLinkedQueue<>();
+        this.listeners = new HashMap<>();
+        this.listenerQueue = new LinkedList<>();
```
Comment (AlbumenJ) [inline dubbo-registry/dubbo-registry-api/src/main/java/org/apache/dubbo/registry/client/event/listener/ServiceInstancesChangedListener.java:86]: `listenerQueue` will be used concurrency in `notifyAddressChanged` method.

**Review by AlbumenJ — CHANGES_REQUESTED** [review]

File: dubbo-registry/dubbo-registry-api/src/main/java/org/apache/dubbo/registry/client/event/listener/ServiceInstancesChangedListener.java (hunk 2 above)
Comment (BurningCN) [reply dubbo-registry/dubbo-registry-api/src/main/java/org/apache/dubbo/registry/client/event/listener/ServiceInstancesChangedListener.java:86]: Currently the `notifyAddressChanged` method is only called by the `doOnEvent` method with the `synchronized` modifier. The concurrent scenario you are talking about is that `notifyAddressChanged` may be called by subclasses in the future?

**Comment by chickenlj** [comment]
Thanks for the contribution. The implementation in the main repo has already been refactored to meet some requirements, I will close this one now

**Pull request closed without merging**